### PR TITLE
fix(build_palette): fix broken reload in stateless palette;

### PR DESCRIPTION
**Pull request by 5-pebbles**
So somehow I missed that `C = vim.deepcopy(P)` (#125 ) would break the upvalue reference, then somehow I did not notice that while testing (If I did any testing, that is, I don't really remember...).

Anyway, this should fix that, and it's more readable now.

Super-Duper, sorry about this (´°ω°`)

### CODE REVIEW

File: lua/nordic/colors/init.lua [hunk 1]
```diff
@@ -3,13 +3,17 @@ local P = require 'nordic.colors.nordic'
...
+    -- clear all values except build_palette
+    for k, _ in pairs(C) do
+        if k ~= "build_palette" then
+            C[k] = nil
+        end
+    end
+    -- copy all values from the base palette
+    for k, v in pairs(vim.deepcopy(P)) do
+        C[k] = v
+    end
```
Comment (AlexvZyl) [inline lua/nordic/colors/init.lua:16]: Maybe we should move all of this out into a function?

**Comment by AlexvZyl** [comment]
Super chill, this is why we have a dev branch :)

File: lua/nordic/colors/init.lua (hunk 1 above)
Comment (5-pebbles) [reply lua/nordic/colors/init.lua]: I did that when I merged it with `changing_palette`, but I didn't think of that here I will update it

File: lua/nordic/colors/init.lua [hunk 2]
```diff
@@ -102,7 +104,6 @@ local function build_palette()
 end
 
 -- build the first palette
-build_palette()
--- after this we can call it with C.build_palette()
+C.build_palette()
```
[the remark this answers is not in the document]
Comment (5-pebbles) [reply lua/nordic/colors/init.lua]: Nope its not: [refactor(cleanup): rm useless call of build_palette;](https://github.com/AlexvZyl/nordic.nvim/pull/137/commits/82dc16c6d2dbe139c0723ae3173e71ef1024d444)

**Comment by 5-pebbles** [comment]
I converted it to a draft to make sure it does not get merged yet (I should use this feature more...)

So here's the issue:

```lua
    -- Modify the palette before generating colors.
    C = options.on_palette(C)
```

that works fine...

I think its just the same palette so even though you are reassigning `C` its just going back to the same ref...

Which would probably mean that merge should work?

The only thing I know does not work is `C = vim.deepcopy(P)`. I will figure it out when I get the chance

**Comment by 5-pebbles** [comment]
So I check the neovim source & `tbl_deep_extend` does create a new table: 

```lua
function vim.tbl_deep_extend(behavior, ...)
  return tbl_extend(behavior, true, ...)
end
```

```lua
local function tbl_extend(behavior, deep_extend, ...)
  if behavior ~= 'error' and behavior ~= 'keep' and behavior ~= 'force' then
    error('invalid "behavior": ' .. tostring(behavior))
  end

  if select('#', ...) < 2 then
    error(
      'wrong number of arguments (given '
        .. tostring(1 + select('#', ...))
        .. ', expected at least 3)'
    )
  end

  -- it creates a new table dang \/
  local ret = {} --- @type table<any,any>
  
  if vim._empty_dict_mt ~= nil and getmetatable(select(1, ...)) == vim._empty_dict_mt then
    ret = vim.empty_dict()
  end

  for i = 1, select('#', ...) do
    local tbl = select(i, ...)
    vim.validate({ ['after the second argument'] = { tbl, 't' } })
    --- @cast tbl table<any,any>
    if tbl then
      for k, v in pairs(tbl) do
        if deep_extend and can_merge(v) and can_merge(ret[k]) then
          ret[k] = tbl_extend(behavior, true, ret[k], v)
        elseif behavior ~= 'force' and ret[k] ~= nil then
          if behavior == 'error' then
            error('key found in more than one map: ' .. k)
          end
        else
          ret[k] = v
        end
      end
    end
  end
  return ret
end
```

Also, `merge_inplace` can return the table, and things still work fine with `C = U.merge_inplace(C, vim.deepcopy(P))` as long as it's the same ref...

This will cause an **issue** if a user returns a new table in `on_palette` or `after_palette`.
I think this is why [folke/tokyonight.nvim](https://github.com/folke/tokyonight.nvim/tree/main)'s callbacks don't return the result and instead rely on the user modifying the reference.

I feel like we should do the same thing. What do you think?

> **Note:** this shouldn't break configs which return the result

**Comment by AlexvZyl** [comment]
> I feel like we should do the same thing. What do you think?
> 
> > **Note:** this shouldn't break configs which return the result

You are right.  We should do this.

**Comment by 5-pebbles** [comment]
I will leave this a draft for a day or two just to make sure everything is working, but it should be good to go :)

> Also there was some other issue, but I can't remember what or which branch it was on for the life of me.

File: lua/nordic/colors/init.lua [hunk 3]
```diff
@@ -3,13 +3,15 @@ local P = require 'nordic.colors.nordic'
...
-    -- Because of the override we need to re-add the build_palette to the C table.
-    C.build_palette = build_palette
-    -- If we had attached build_palette to C we would lost it after the override.
+function C.build_palette()
+    -- clear all values except build_palette
+    for k, _ in pairs(C) do
+        if k ~= "build_palette" then
+            C[k] = nil
+        end
+    end
```
Comment (AlexvZyl) [inline lua/nordic/colors/init.lua:12]: I can't remember, but why do we have to clear all of the values before merging?  Can't we just merge in place directly?

File: lua/nordic/colors/init.lua [hunk 4]
```diff
@@ -3,13 +3,15 @@ local P = require 'nordic.colors.nordic'
...
+    -- copy all values from the base palette
+    U.merge_inplace(C, vim.deepcopy(P))
```
Comment (AlexvZyl) [inline lua/nordic/colors/init.lua:14]: Are you sure we have to deepcopy?

Comment (5-pebbles) [reply lua/nordic/colors/init.lua:14]: Nope I will look into that

Comment (5-pebbles) [reply lua/nordic/colors/init.lua]: Not if I change `merge_inplace` to a deep merge:

[feat(utils)!: merge_inplace now deep merges;](https://github.com/AlexvZyl/nordic.nvim/pull/137/commits/dd23021ea48defb129a4e9360d7669b04dbd2a67)
[refactor(colors): clean up build_palette;](https://github.com/AlexvZyl/nordic.nvim/pull/137/commits/4c697537d020fa2df54c89e3d04f9ad8cc542579)

File: lua/nordic/colors/init.lua (hunk 3 above)
Comment (5-pebbles) [reply lua/nordic/colors/init.lua]: I don't remember either:
[refactor(colors): clean up build_palette;](https://github.com/AlexvZyl/nordic.nvim/pull/137/commits/4c697537d020fa2df54c89e3d04f9ad8cc542579)

File: lua/nordic/colors/init.lua (hunk 4 above)
Comment (AlexvZyl) [reply lua/nordic/colors/init.lua]: Nice.

**Review by AlexvZyl — APPROVED** [review]

**Comment by AlexvZyl** [comment]
I think we need some tests for our utils, but I will add an issue for that.

File: lua/nordic/colors/init.lua (hunk 2 above)
[the remark this answers is not in the document]
Comment (5-pebbles) [reply lua/nordic/colors/init.lua]: It turns out that this is necessary if the user imports `nordic.colors` in their config...

I will just sneak the fix into the next PR.

**Pull request merged**